### ISSUE TITLE
Regression in Piecewise result

### ISSUE THREAD

**Issue by mgeier** [sympy/sympy#22747]
I might be doing something wrong, but with the latest `master` branch I'm getting a different result than previously.

The example with more details is available at https://splines.readthedocs.io/en/0.1.0/euclidean/catmull-rom-uniform.html#Cardinal-Functions.

Here's just a small reproducer (which could for sure be made even more minimal):

```python
import sympy as sp

t = sp.symbols('t')
i, j, k = sp.symbols('i j k', integer=True)
w = sp.Function('w')
C0k = sp.Sum(
        sp.Product(
            sp.Piecewise((1, sp.Eq(j, 0)), ((t / j) + 1, True)),
            (j, i - k, i)) * w(t + i),
        (i, 0, k))
B0 = sp.Piecewise((0, t < i), (1, t < i + 1), (0, True))
B1 = (t - i) * B0 + (i + 2 - t) * B0.subs(i, i + 1)
B2 = (t - i) / 2 * B1 + (i + 3 - t) / 2 * B1.subs(i, i + 1)
w2 = B2.subs(i, -1)
C01 = C0k.subs(k, 1).replace(w, lambda x: w2.subs(t, x)).doit().simplify()
sp.printing.pprint(C01)
```

### SymPy 1.9 (and before)

```
⎧        0          for t < -2
⎪                             
⎪               2             
⎪(t + 1)⋅(t + 2)              
⎪────────────────   for t < -1
⎪       2                     
⎪                             
⎪     3      2                
⎪  3⋅t    5⋅t                 
⎪- ──── - ──── + 1  for t < 0 
⎪   2      2                  
⎨                             
⎪    3      2                 
⎪ 3⋅t    5⋅t                  
⎪ ──── - ──── + 1   for t < 1 
⎪  2      2                   
⎪                             
⎪               2             
⎪(1 - t)⋅(t - 2)              
⎪────────────────   for t < 2 
⎪       2                     
⎪                             
⎩        0          otherwise
```

### `master` branch since 34289ce4d92cc9424b05a01b7fa74d156853c392

```
⎧                  0                         for t < -2    
⎪                                                          
⎪                          2                               
⎪           (t + 1)⋅(t + 2)                                
⎪           ────────────────                 for t < -1    
⎪                  2                                       
⎪                                                          
⎪                                 2                        
⎪   3      2   t   (1 - t)⋅(t + 1)    1                    
⎪- t  - 2⋅t  - ─ + ──────────────── + ─  for t ≤ -2 ∨ t < 0
⎪              2          2           2                    
⎪                                                          
⎪                      2                                   
⎨               1 - 4⋅t                      for t ≤ -1    
⎪                                                          
⎪                        2                                 
⎪  3      2   t   (t - 1) ⋅(t + 1)   1                     
⎪ t  - 2⋅t  + ─ + ──────────────── + ─       for t < 1     
⎪             2          2           2                     
⎪                                                          
⎪                          2                               
⎪           (1 - t)⋅(t - 2)                                
⎪           ────────────────                 for t < 2     
⎪                  2                                       
⎪                                                          
⎪                  0                         otherwise     
⎩                                                          
```

Not only is this different, but the ranges also look kinda dubious, don't they?

The regression was introduced in #22286 by @smichr.

**Comment by oscarbenjamin** [comment]
CC @smichr 

I believe that these are equivalent although the previous result was simpler.

1. The first and second cases of both are the same.
2. The third cases are equivalent if you expand the brackets and recognise that `t <= -2` is superfluous in combination with `| t < 0`.
3. The 4th case in the master output is impossible (we can't have `t <= -1` if the 3rd case `t < 0` didn't hold).
4. The remaining cases are equivalent if you expand/factor them.

So I think that semantically the two Piecewise are equivalent.

The release note from #22286 says:
```
Piecewise.eval is faster: removed steps that should be in simplification
```
So I guess the intention is that to get previous results you would use simplify, expand, factor etc but `simplify` doesn't change this output:
```python
In [76]: C01.simplify() == C01
Out[76]: True
```

**Comment by smichr** [comment]
@oscargus has some work related to simplification of Piecewise, I believe. This is something that you might want to put in your library:
```python
def cascade(p, x):
    """rewrite conditions of piecewise in increasing fashion if possible.

    Examples
    ========

    >>> from sympy import Piecewise
    >>> from smypy.abc import t
    >>> p = Piecewise(
    ...     (0, t < -2),
    ...     ((t + 1)*(t + 2)**2/2, t < -1),
    ...     (-t**3 - 2*t**2 - t/2 + (1 - t)*(t + 1)**2/2 + 1/2, (t <= -2) | (t < 0)),
    ...     (1 - 4*t**2, t <= -1),
    ...     (t**3 - 2*t**2 + t/2 + (t - 1)**2*(t + 1)/2 + 1/2, t < 1),
    ...     ((1 - t)*(t - 2)**2/2, t < 2),
    ...     (0, True))
    >>> cascade(p, t)
    Piecewise(
        (0, t < -2),
        ((t + 1)*(t + 2)**2/2, t < -1),
        (-3*t**3/2 - 5*t**2/2 + 1, t < 0),
        (3*t**3/2 - 5*t**2/2 + 1, t < 1),
        ((1 - t)*(t - 2)**2/2, t < 2),
        (0, True))
    """
    ok, iv = p._intervals(x)
    if not ok:
        print('intervals gave False for first arg')
        return p
    sargs = []
    cov = S.EmptySet
    first = True
    for a,b,e,i in iv:
      if first and a != S.NegativeInfinity:
          print('not a Piecewise covering from -oo')
          return p
      s = Interval(a, b) - cov
      if s is S.EmptySet:
          continue
      if not isinstance(s, Interval):
          print('unhandled set',s)
          return p
      if p.args[i].cond.subs(x, s.sup) == True:
          c = (x <= s.sup) if s.sup.is_finite else True
      else:
          c = (x < s.sup)
      cov |= s
      sargs.append((simplify(e), c))
      if c == True:
          break
    if len(sargs) == 1 and sargs[0][1] == True:
        return sargs[0][0]
    return p.func(*sargs, evaluate=False)
```
This will restore the output for your expression: `cascade(C01, t)`.

**Comment by smichr** [comment]
I had chance to check that this is mostly already in place. Corrections and improved handling has been submitted as #22751

**Comment by oscarbenjamin** [comment]
To be clear since this issue was automatically closed does #22751 restore the previous behaviour or does it fix it so that `simplify` works in this case?

**Comment by mgeier** [comment]
I don't really understand the fix, but it fixes the regression that I observed.
With the latest `master`, the behavior of SymPy 1.9 is restored, no additional `simplify()` necessary.

Thank you all for the quick fix!
